### ISSUE TITLE
Incorrect inference off of `isinstance` and list

### ISSUE THREAD

**Issue by Dr-Irv** [microsoft/pyright#4302]
**Describe the bug**
```python 
from typing import List, Union


class Base:
    ...


class Ct:
    def __init__(self, rhs: Union[List[int], int, Base, List[Base]]):
        if isinstance(rhs, Base):
            rhs = [rhs]
        self.rhs: Union[List[int], int, List[Base]] = rhs

    def brhs(self):
        if isinstance(self.rhs, Base):
            self.rhs = [self.rhs]
```
pyright reports:
```text
listobj.py
  c:\Code\Misc\pytypes\listobj.py:16:25 - error: Cannot assign member "rhs" for type "Ct"
    Type "<subclass of list and Base> | <subclass of int and Base>" cannot be assigned to type "int"
      "<subclass of list and Base>" is incompatible with "int" (reportGeneralTypeIssues)
```
The `isinstance` test should tell pyright to infer that `self.rhs` is of type `Base`, so that the expression `[self.rhs]` is then `List[Base]`

**Expected behavior**
No error reported

**VS Code extension or command-line**

pyright 1.1.283 command line, python 3.9

**Comment by Dr-Irv** [comment]
`mypy` does not report an error here.

You're right that the line `self.rhs = [self.rhs]` is dead code.  This was 3 year old code, and pyright only recently picked it up.  But I disagree with your analysis. 

`self.rhs` is typed as `int | List[int] | List[Base]`.  So `isinstance(self.rhs, Base)` will always be False.  The `isinstance()` check can't be true.  But I still don't think an error should be reported, and the code should be greyed out.

**Comment by erictraut** [comment]
Mypy doesn't analyze methods that are unannotated. If you add a `-> None` return type annotation to that method, it will then analyze it and output the same error as pyright.

> So isinstance(self.rhs, Base) will always be False.

No, that's not right. You could have a type like `class Foo(List[int], Base): ...` that is both a subclass of `List[int]` and `Base`.

**Comment by Dr-Irv** [comment]
OK, I got it.  I see that it is inferring that it would have to be a subclass that had 2 parents.
